### ISSUE TITLE
Possibility of using built-in Mahalanobis distance

### ISSUE THREAD

**Issue by aazuspan** [lemma-osu/scikit-learn-knn-regression#38]
Hey @grovduck, I was poking around sklearn and came across a built-in implementation of Mahalanobis as a [distance metric](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.DistanceMetric.html). After reading through some [stackoverflow discussions](https://stackoverflow.com/questions/34643548/how-to-use-mahalanobis-distance-in-sklearn-distancemetrics/34650347#34650347), I was able to bodge together an alternate implementation of `MahalanobisKNNRegressor` that doesn't require a custom transformer and produces (as far as I can tell) identical results.

```python
class MahalanobisKNNRegressor(IDNeighborsRegressor):
    def fit(self, X, y):
        self.metric = "mahalanobis"
        self.metric_params = {"VI": np.linalg.inv(np.cov(X, rowvar=False))}
        return super().fit(X, y)
```

This implementation is pretty rough and I don't really understand how it works, but it does pass all tests. There are definitely upsides to this simpler approach in that we can remove some code that's ported from `yaImpute` and cut down on the number of estimator checks we need to pass, but do you think we're losing anything in terms of future flexibility? Or is there additional value to including the `MahalanobisTransformer` beyond its use in this estimator? Curious to hear your thoughts!

**Comment by grovduck** [comment]
Hey @aazuspan, great find!  Definitely more successful than my wrangling with the sklearn `CCA`.  I don't *think* there is any reason to hold onto the transformer if this works for the regressor.  The only lingering doubt is that we're currently relying on `self.transform_` in EUC, MAH, GNN, MSN to provide the transformation to get to strict Euclidean neighbor finding (RAW is already there), so we'd be taking Mahalanobis out of this paradigm, right?  I don't think that's a problem because distance spaces aren't necessarily comparable.  

I think the one place it may come up is with the idea of creating "synthetic" plots (we talked about this very early on), where we first perform the ordination on the plots to put them in terms of the transformed space and then lay down a lattice of points (the synthetic plots) based on equal interval or quantile spacing within that space.  This wouldn't necessarily happen within `sknnr` and we've also only created the synthetic plots in `CCA` space, so again not a show stopper.  

What do you think based on these possible future scenarios?

**Comment by aazuspan** [comment]
> The only lingering doubt is that we're currently relying on self.transform_ in EUC, MAH, GNN, MSN to provide the transformation to get to strict Euclidean neighbor finding (RAW is already there), so we'd be taking Mahalanobis out of this paradigm, right?

Yeah, I think this is a valid concern. Even though it simplifies the code, this probably adds some conceptual complexity since this estimator would work differently from the others.

> I think the one place it may come up is with the idea of creating "synthetic" plots (we talked about this very early on), where we first perform the ordination on the plots to put them in terms of the transformed space and then lay down a lattice of points (the synthetic plots) based on equal interval or quantile spacing within that space. This wouldn't necessarily happen within sknnr and we've also only created the synthetic plots in CCA space, so again not a show stopper.

Good point. Even if that functionality isn't included in `sknnr`, I assume we'd need a Mahalanobis transformer *somewhere* if we ever wanted to generate synthetic plots in that space?

> What do you think based on these possible future scenarios?

Given that we already have a working solution and we could *potentially* be shooting ourselves in the foot with this, I think I lean towards shelving this for now, and revisiting if we ever change our minds in the future. Let me know if you disagree, but otherwise I think we can close this!

**Comment by grovduck** [comment]
@aazuspan, I hate to waste the effort that you put into it, but I think this is probably the right decision at this point.  Luckily, it's not a terribly complex estimator and we'll have the two implementation methods ready to do if we need them.  Thanks for looking into this and suggesting it.
